src/lib/device-online-state/index.ts

```diff
@@ -0,0 +1,280 @@
+import * as Bluebird from 'bluebird';
+import { EventEmitter } from 'events';
+import { DEFAULT_SUPERVISOR_POLL_INTERVAL } from '../env-vars';
+import { once, noop } from 'lodash';
+import { resinApi, root } from '../../platform';
+import { captureException } from '../../platform/errors';
+import {
+	createPromisifedRedisClient,
+	PromisifedRedisClient,
+} from '../redis-promise';
+import * as RedisSMQ from 'rsmq-promise';
+
+export class PollIntervalNotFoundError extends Error {}
+export const getPollIntervalForDevice = (uuid: string): Bluebird<number> => {
+	return resinApi
+		.get({
+			resource: 'device_config_variable',
+			passthrough: {
+				req: root,
+			},
+			options: {
+				$select: 'value',
+				$expand: {
+					device: {
+						$filter: { uuid },
+					},
+				},
+				$filter: {
+					name: 'RESIN_SUPERVISOR_POLL_INTERVAL',
+				},
+			},
+		})
+		.then(([pollInterval]: Array<{ value: number }>) => {
+			if (!pollInterval) {
+				throw new PollIntervalNotFoundError();
+			}
+
+			return pollInterval.value;
+		})
+		.catch(PollIntervalNotFoundError, () => {
+			// query to determine the poll interval for this device...
+			return resinApi
+				.get({
+					resource: 'application_config_variable',
+					passthrough: {
+						req: root,
+					},
+					options: {
+						$select: 'value',
+						$top: 1,
+						$filter: {
+							application: {
+								$any: {
+									$alias: 'a',
+									$expr: {
+										a: {
+											owns__device: {
+												$any: {
+													$alias: 'd',
+													$expr: {
+														d: {
+															uuid: uuid,
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+							name: 'RESIN_SUPERVISOR_POLL_INTERVAL',
+						},
+					},
+				})
+				.then(([pollInterval]: Array<{ value: number }>) => {
+					if (pollInterval) {
+						return pollInterval.value;
+					}
+
+					return DEFAULT_SUPERVISOR_POLL_INTERVAL;
+				});
+		});
+};
+
+export const getDefaultManager = once(() => {
+	return new DeviceOnlineStateManager();
+});
+
+export enum DeviceOnlineStates {
+	Unknown = -2,
+	Timeout = -1,
+	Offline = 0,
+	Online = 1,
+}
+
+class NoMessagesToConsumeSignal extends Error {}
+
+class DeviceOnlineStateManager extends EventEmitter {
+	private static readonly REDIS_NAMESPACE = 'device-online-state';
+	private static readonly EXPIRED_QUEUE = 'expired';
+
+	rsmq: RedisSMQ;
+	redis: PromisifedRedisClient;
+
+	constructor() {
+		super();
+
+		// create a new Redis client...
+		this.redis = createPromisifedRedisClient({
+			host: process.env.REDIS_HOST,
+			port: parseInt(process.env.REDIS_PORT || '6379'),
```

Comment:
This should be fetched/defaulted in the config module (all process.env accesses should be there)

Reply:
They were already, I just forgot to use them :grin: Done